src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -439,6 +456,243 @@ instance Function Address where
+          -- For the outer shrink, we can allow empty prefixes
+          xs' :: [(CurrencySymbol, [(TokenName, Integer)])] <-
+            init . subsequences $ xs >>= \entry -> traverse shrinkSubseq entry
+          -- Glue these together, then rewrap
+          pure . PlutusTx.fromList . fmap (fmap PlutusTx.fromList) $ x' : xs'
+
+-- TODO: Enable zero ADA 'Value' generation.
+
+-- | @since 2.1.3
+instance CoArbitrary Value where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary (Value inner) =
+    coarbitrary (fmap (fmap PlutusTx.toList) . PlutusTx.toList $ inner)
+
+-- | @since 2.1.3
+instance Function Value where
+  {-# INLINEABLE function #-}
+  function = functionMap into $ \inner ->
+    Value . AssocMap.fromList . fmap (fmap AssocMap.fromList) $ inner
+    where
+      into :: Value -> [(CurrencySymbol, [(TokenName, Integer)])]
+      into (Value inner) =
+        fmap (fmap PlutusTx.toList) . PlutusTx.toList $ inner
+
+{- | The shrinker for this type will not shrink \'out-of-arm\'. Effectively,
+ this means 'NoOutputDatum' does not shrink, 'OutputDatumHash' also does not
+ shrink (as 'DatumHash' doesn't), and 'OutputDatum' will shrink to other
+ 'OutputDatum's.
+
+ @since 2.1.3
+-}
+instance Arbitrary OutputDatum where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary =
+    Gen.oneof
+      [ pure NoOutputDatum
+      , OutputDatumHash <$> arbitrary
+      , OutputDatum <$> arbitrary
+      ]
+  {-# INLINEABLE shrink #-}
+  shrink = \case
+    NoOutputDatum -> []
+    OutputDatumHash _ -> []
+    OutputDatum dat -> OutputDatum <$> shrink dat
+
+-- | @since 2.1.3
+instance CoArbitrary OutputDatum where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary od = case od of
+    NoOutputDatum -> Gen.variant (0 :: Int)
+    OutputDatumHash dh -> Gen.variant (1 :: Int) . coarbitrary dh
+    OutputDatum dat -> Gen.variant (2 :: Int) . coarbitrary dat
+
+-- | @since 2.1.3
+instance Function OutputDatum where
+  {-# INLINEABLE function #-}
+  function = functionMap into $ \case
+    Nothing -> NoOutputDatum
+    Just (Left dh) -> OutputDatumHash dh
+    Just (Right dat) -> OutputDatum dat
+    where
+      into :: OutputDatum -> Maybe (Either DatumHash Datum)
+      into = \case
+        NoOutputDatum -> Nothing
+        OutputDatumHash dh -> Just . Left $ dh
+        OutputDatum dat -> Just . Right $ dat
+
+{- | This is based on the documentation of 'ScriptHash', specifically that it
+ represents a SHA256 hash. In particular, this type does not shrink, as it
+ wouldn't really make much sense to.
+
+ @since 2.1.3
+-}
+deriving via DatumHash instance Arbitrary ScriptHash
+
+-- | @since 2.1.3
+deriving via DatumHash instance CoArbitrary ScriptHash
+
+-- | @since 2.1.3
+instance Function ScriptHash where
+  {-# INLINEABLE function #-}
+  function = functionMap (coerce @_ @BuiltinByteString) coerce
+
+-- | @since 2.1.3
+instance Arbitrary TxOut where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = do
+    addr <- arbitrary
```

Comment:
```hs
TxOut <$> arbitrary <*> arbitrary <*> arbitrary <*> arbitrary
```
What do you think?

Reply:
The main reason I didn't do this is because a field reordering upstream would break it. I don't object to changing to the applicative style instead though.

Reply:
I'm not entirely opposed to explicitly listing the fields like this in a do block